File: app/src/main/java/com/intelligentquestionbank/memory/database/dao/StudyRecordDao.java
package com.intelligentquestionbank.memory.database.dao;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;
import com.intelligentquestionbank.memory.database.entities.StudyRecord;
import java.util.List;

@Dao
public interface StudyRecordDao {
    
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    long insert(StudyRecord studyRecord);
    
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertAll(StudyRecord... studyRecords);
    
    @Update
    void update(StudyRecord studyRecord);
    
    @Delete
    void delete(StudyRecord studyRecord);
    
    @Query("DELETE FROM study_records WHERE id = :recordId")
    void deleteById(long recordId);
    
    @Query("SELECT * FROM study_records WHERE question_id = :questionId")
    StudyRecord getStudyRecordByQuestionId(long questionId);
    
    @Query("SELECT * FROM study_records WHERE question_id = :questionId")
    LiveData<StudyRecord> getStudyRecordByQuestionIdLive(long questionId);
    
    @Query("SELECT * FROM study_records")
    List<StudyRecord> getAllStudyRecords();
    
    @Query("SELECT * FROM study_records")
    LiveData<List<StudyRecord>> getAllStudyRecordsLive();
    
    @Query("SELECT * FROM study_records WHERE is_mastered = 0 ORDER BY next_review_time ASC")
    List<StudyRecord> getActiveStudyRecords();
    
    @Query("SELECT * FROM study_records WHERE is_error = 1")
    List<StudyRecord> getErrorStudyRecords();
    
    @Query("SELECT COUNT(*) FROM study_records WHERE is_mastered = 1")
    LiveData<Integer> getMasteredCount();
    
    @Query("SELECT COUNT(*) FROM study_records WHERE is_error = 1")
    LiveData<Integer> getErrorCount();
    
    @Query("SELECT AVG(CAST(correct_attempts AS REAL) / total_attempts * 100) FROM study_records WHERE total_attempts > 0")
    LiveData<Double> getAverageCorrectRate();
    
    @Query("UPDATE study_records SET is_error = :isError WHERE question_id = :questionId")
    void updateErrorStatus(long questionId, boolean isError);
    
    @Query("UPDATE study_records SET is_mastered = :isMastered WHERE question_id = :questionId")
    void updateMasteredStatus(long questionId, boolean isMastered);
    
    @Query("DELETE FROM study_records WHERE question_id IN " +
           "(SELECT id FROM questions WHERE library_id = :libraryId)")
    void deleteRecordsByLibraryId(long libraryId);
    
    @Query("DELETE FROM study_records")
    void deleteAll();
}

File: app/src/main/java/com/intelligentquiz/memory/database/dao/StudyRecordDao.kt
package com.intelligentquiz.memory.database.dao

import androidx.room.*
import androidx.lifecycle.LiveData
import com.intelligentquiz.memory.database.entities.StudyRecord
import kotlinx.coroutines.flow.Flow

@Dao
interface StudyRecordDao {
    
    @Query("SELECT * FROM study_records WHERE questionId = :questionId ORDER BY studiedAt DESC")
    fun getRecordsByQuestion(questionId: Long): Flow<List<StudyRecord>>
    
    @Query("SELECT * FROM study_records WHERE questionId = :questionId ORDER BY studiedAt DESC LIMIT 1")
    suspend fun getLatestRecord(questionId: Long): StudyRecord?
    
    @Insert
    suspend fun insertRecord(record: StudyRecord): Long
    
    @Update
    suspend fun updateRecord(record: StudyRecord)
    
    @Query("SELECT COUNT(*) FROM study_records sr JOIN questions q ON sr.questionId = q.id WHERE q.libraryId = :libraryId AND sr.consecutiveCorrect >= 3")
    suspend fun getMasteredQuestionCount(libraryId: Long): Int
    
    @Query("SELECT COUNT(*) FROM study_records sr JOIN questions q ON sr.questionId = q.id WHERE q.libraryId = :libraryId AND sr.isInErrorBank = 1")
    suspend fun getErrorBankCount(libraryId: Long): Int
    
    @Query("SELECT COUNT(*) FROM study_records sr JOIN questions q ON sr.questionId = q.id WHERE q.libraryId = :libraryId AND DATE(sr.studiedAt/1000, 'unixepoch') = DATE('now')")
    suspend fun getTodayStudyCount(libraryId: Long): Int
    
    @Query("SELECT COUNT(*) FROM study_records sr JOIN questions q ON sr.questionId = q.id WHERE q.libraryId = :libraryId AND sr.remembered = 1")
    suspend fun getCorrectAnswerCount(libraryId: Long): Int
    
    @Query("SELECT COUNT(*) FROM study_records sr JOIN questions q ON sr.questionId = q.id WHERE q.libraryId = :libraryId")
    suspend fun getTotalStudyCount(libraryId: Long): Int
    
    @Query("SELECT AVG(sr.responseTime) FROM study_records sr JOIN questions q ON sr.questionId = q.id WHERE q.libraryId = :libraryId AND sr.responseTime > 0")
    suspend fun getAverageResponseTime(libraryId: Long): Double?
    
    @Query("DELETE FROM study_records WHERE questionId = :questionId")
    suspend fun deleteRecordsByQuestion(questionId: Long)
    
    @Query("SELECT * FROM study_records sr JOIN questions q ON sr.questionId = q.id WHERE q.libraryId = :libraryId AND sr.nextReviewAt <= :currentTime ORDER BY sr.nextReviewAt")
    suspend fun getQuestionsForReview(libraryId: Long, currentTime: Long): List<StudyRecord>
}